node/src/components/deploy_buffer.rs

```diff
@@ -76,26 +80,74 @@ impl Display for Event {
     }
 }
 
+type DeployCollection = HashMap<DeployHash, DeployHeader>;
+type ProtoBlockCollection = HashMap<ProtoBlockHash, DeployCollection>;
+/// Deploy buffer.
+#[derive(DataSize, Debug, Default, Clone)]
+struct DeployBufferInner {
+    block_max_deploy_count: usize,
+    collected_deploys: DeployCollection,
+    processed: ProtoBlockCollection,
+    finalized: ProtoBlockCollection,
+}
+
 /// Deploy buffer.
 #[derive(DataSize, Debug, Clone)]
 pub(crate) struct DeployBuffer {
-    block_max_deploy_count: usize,
-    collected_deploys: HashMap<DeployHash, DeployHeader>,
-    processed: HashMap<ProtoBlockHash, HashMap<DeployHash, DeployHeader>>,
-    finalized: HashMap<ProtoBlockHash, HashMap<DeployHash, DeployHeader>>,
+    inner: Arc<Mutex<DeployBufferInner>>, // Don't hand out references to internal mutable state
```

Comment:
I'm not a fan of the Arc<Mutex<..>> approach taken here. Did you consider any alternatives? At least one other method takes a `&mut self,`, and also afaik all incoming activity is received from a single reactor at runtime.

Reply:
Yep I'm rewriting to make use of the reactor and events to do this.